File: ghost_monitor/market_impact.py
from __future__ import annotations

import json
import math
import sys
import time
import urllib.parse
import urllib.request
from datetime import date, datetime, timedelta
from pathlib import Path
from typing import Any

from .qveris_client import ssl_context

ROOT = Path(__file__).resolve().parents[1]
PRICE_CACHE = ROOT / "data" / "price_window_cache.json"
_CACHE: dict[str, Any] | None = None


def _load_cache() -> dict[str, Any]:
    global _CACHE
    if _CACHE is None:
        _CACHE = json.loads(PRICE_CACHE.read_text()) if PRICE_CACHE.exists() else {}
    return _CACHE


def _save_cache(cache: dict[str, Any]) -> None:
    PRICE_CACHE.parent.mkdir(exist_ok=True)
    PRICE_CACHE.write_text(json.dumps(cache, indent=2, ensure_ascii=False))


def yahoo_symbol(symbol: str) -> str:
    if symbol.endswith(".KS"):
        return symbol
    return symbol


def fetch_chart(symbol: str, start: date, end: date) -> list[dict[str, Any]]:
    key = f"{symbol}|{start.isoformat()}|{end.isoformat()}"
    cache = _load_cache()
    if key in cache and not stale_cache(cache[key], end):
        return cache[key]
    p1 = int(time.mktime(start.timetuple()))
    p2 = int(time.mktime((end + timedelta(days=1)).timetuple()))
    encoded = urllib.parse.quote(yahoo_symbol(symbol), safe="")
    url = f"https://query1.finance.yahoo.com/v8/finance/chart/{encoded}?period1={p1}&period2={p2}&interval=1d&events=history"
    req = urllib.request.Request(url, headers={"user-agent": "ghost-monitor/0.1"})
    with urllib.request.urlopen(req, timeout=20, context=ssl_context()) as resp:
        payload = json.loads(resp.read())
    result = payload.get("chart", {}).get("result") or []
    if not result:
        return []
    rows = []
    timestamps = result[0].get("timestamp") or []
    quote = (result[0].get("indicators", {}).get("quote") or [{}])[0]
    adj = (result[0].get("indicators", {}).get("adjclose") or [{}])[0].get("adjclose") or quote.get("close") or []
    for i, ts in enumerate(timestamps):
        close = adj[i] if i < len(adj) else None
        volume = (quote.get("volume") or [None])[i]
        if close is None or (isinstance(close, float) and math.isnan(close)):
            continue
        rows.append({"date": datetime.utcfromtimestamp(ts).date().isoformat(), "close": float(close), "volume": volume})
    cache[key] = rows
    _save_cache(cache)
    return rows


def latest_available_trading_day(now_utc: datetime | None = None) -> date:
    now_utc = now_utc or datetime.utcnow()
    day = now_utc.date()
    if not is_us_trading_day(day):
        return previous_trading_day(day)
    # US cash close is 20:00 UTC during daylight saving time. The extra hour
    # gives Yahoo's daily bar time to settle before we treat it as available.
    return day if now_utc.hour >= 21 else previous_trading_day(day - timedelta(days=1))


def stale_cache(rows: list[dict[str, Any]], end: date) -> bool:
    if not rows:
        return False
    required = previous_trading_day(min(end, latest_available_trading_day()))
    return rows[-1]["date"] < required.isoformat()


def first_on_or_after(rows: list[dict[str, Any]], event_date: str, offset: int = 0) -> dict[str, Any] | None:
    eligible = [r for r in rows if r["date"] >= event_date]
    if len(eligible) <= offset:
        return None
    return eligible[offset]


def last_before(rows: list[dict[str, Any]], event_date: str) -> dict[str, Any] | None:
    eligible = [r for r in rows if r["date"] < event_date]
    return eligible[-1] if eligible else None


def pct(a: float, b: float) -> float:
    return round((b / a - 1) * 100, 2)


def parse_event_at(value: str) -> datetime:
    raw = str(value or "").replace("Z", "+00:00")
    if len(raw) >= 8 and raw[:8].isdigit() and (len(raw) == 8 or raw[8] == "T"):
        tail = "".join(ch for ch in raw[9:] if ch.isdigit())[:6].ljust(6, "0")
        return datetime.strptime(raw[:8] + tail, "%Y%m%d%H%M%S")
    return datetime.fromisoformat(raw if "T" in raw else f"{raw}T00:00:00")


US_MARKET_HOLIDAYS = {
    # NYSE/Nasdaq full-day closures. Keep this small and explicit for the
    # current backfill window instead of pulling in a calendar dependency.
    "2024-01-01", "2024-01-15", "2024-02-19", "2024-03-29", "2024-05-27",
    "2024-06-19", "2024-07-04", "2024-09-02", "2024-11-28", "2024-12-25",
    "2025-01-01", "2025-01-09", "2025-01-20", "2025-02-17", "2025-04-18",
    "2025-05-26", "2025-06-19", "2025-07-04", "2025-09-01", "2025-11-27",
    "2025-12-25",
    "2026-01-01", "2026-01-19", "2026-02-16", "2026-04-03", "2026-05-25",
    "2026-06-19", "2026-07-03", "2026-09-07", "2026-11-26", "2026-12-25",
}


def is_us_trading_day(day: date) -> bool:
    return day.weekday() < 5 and day.isoformat() not in US_MARKET_HOLIDAYS


def next_trading_day(day: date) -> date:
    while not is_us_trading_day(day):
        day += timedelta(days=1)
    return day


def previous_trading_day(day: date) -> date:
    while not is_us_trading_day(day):
        day -= timedelta(days=1)
    return day


def expected_reaction_date(event_at: datetime) -> str:
    day = event_at.date()
    if not is_us_trading_day(day):
        return next_trading_day(day).isoformat()
    # Approximate exchange close in UTC. Daily bars confirm the exact result once available.
    if event_at.hour >= 20:
        return next_trading_day(day + timedelta(days=1)).isoformat()
    return day.isoformat()


def reaction_rows(rows: list[dict[str, Any]], event_at: datetime) -> tuple[dict[str, Any] | None, dict[str, Any] | None, dict[str, Any] | None, dict[str, Any] | None]:
    day = event_at.date().isoformat()
    after_close = event_at.hour >= 20
    eligible = [r for r in rows if r["date"] > day] if after_close else [r for r in rows if r["date"] >= day]
    reaction = eligible[0] if eligible else None
    if not reaction:
        prev = last_before(rows, day)
        return prev, None, None, None
    prev = last_before(rows, reaction["date"])
    next_row = eligible[1] if len(eligible) > 1 else None
    third_row = eligible[3] if len(eligible) > 3 else None
    return prev, reaction, next_row, third_row


def impact_for(symbol: str, event_date: str) -> dict[str, Any]:
    event_at = parse_event_at(event_date)
    start = event_at.date() - timedelta(days=10)
    end = event_at.date() + timedelta(days=10)
    try:
        rows = fetch_chart(symbol, start, end)
        prev, d0, d1, d3 = reaction_rows(rows, event_at)
        if not prev:
            return {"symbol": symbol, "error": "missing price window"}
        if not d0:
            return {
                "symbol": symbol,
                "pending": "waiting_for_reaction_close",
                "expected_reaction_date": expected_reaction_date(event_at),
                "prev_date": prev["date"],
                "prev_close": round(prev["close"], 4),
            }
        return {
            "symbol": symbol,
            "prev_date": prev["date"],
            "reaction_trade_date": d0["date"],
            "prev_close": round(prev["close"], 4),
            "reaction_close": round(d0["close"], 4),
            "next_close": round(d1["close"], 4) if d1 else None,
            "reaction_pct": pct(prev["close"], d0["close"]),
            "next_day_pct": pct(d0["close"], d1["close"]) if d1 else None,
            "three_session_pct": pct(prev["close"], d3["close"]) if d3 else None,
        }
    except Exception as exc:
        return {"symbol": symbol, "error": str(exc)}


def build_report(cases: list[dict[str, Any]]) -> list[dict[str, Any]]:
    out = []
    for case in cases:
        impacts = [impact_for(symbol, case["date"]) for symbol in case["symbols"]]
        out.append({**case, "market_impact": impacts})
    return out


def main(argv: list[str] | None = None) -> int:
    src = Path(argv[0]) if argv else ROOT / "data/historical_cases.json"
    cases = json.loads(src.read_text())
    report = build_report(cases)
    out = ROOT / "reports/historical_cases_with_market_impact.json"
    out.write_text(json.dumps(report, indent=2, ensure_ascii=False))
    print(out)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: app/lib/market-impact.ts
type Row = Record<string, unknown>;

type PriceBar = {
  date: string;
  close: number;
};

type MarketImpact = {
  symbol: string;
  prev_date?: string;
  reaction_trade_date?: string;
  prev_close?: number;
  reaction_close?: number;
  next_close?: number | null;
  reaction_pct?: number;
  next_day_pct?: number | null;
  three_session_pct?: number | null;
  pending?: "waiting_for_reaction_close";
  expected_reaction_date?: string;
  error?: string;
};

const US_MARKET_HOLIDAYS = new Set([
  "2024-01-01", "2024-01-15", "2024-02-19", "2024-03-29", "2024-05-27", "2024-06-19", "2024-07-04", "2024-09-02", "2024-11-28", "2024-12-25",
  "2025-01-01", "2025-01-09", "2025-01-20", "2025-02-17", "2025-04-18", "2025-05-26", "2025-06-19", "2025-07-04", "2025-09-01", "2025-11-27", "2025-12-25",
  "2026-01-01", "2026-01-19", "2026-02-16", "2026-04-03", "2026-05-25", "2026-06-19", "2026-07-03", "2026-09-07", "2026-11-26", "2026-12-25",
]);

const asList = (value: unknown): unknown[] => Array.isArray(value) ? value : [];
const round = (value: number, digits = 2) => Number(value.toFixed(digits));
const pct = (before: number, after: number) => round((after / before - 1) * 100);

function addDays(value: string, days: number) {
  const date = new Date(`${value}T12:00:00Z`);
  date.setUTCDate(date.getUTCDate() + days);
  return date.toISOString().slice(0, 10);
}

export function isUsTradingDay(value: string) {
  const day = new Date(`${value}T12:00:00Z`).getUTCDay();
  return day !== 0 && day !== 6 && !US_MARKET_HOLIDAYS.has(value);
}

function isTradingDay(value: string, symbol = "") {
  if (!symbol.endsWith(".KS")) return isUsTradingDay(value);
  const day = new Date(`${value}T12:00:00Z`).getUTCDay();
  return day !== 0 && day !== 6;
}

function nextTradingDay(value: string, symbol = "") {
  let day = value;
  while (!isTradingDay(day, symbol)) day = addDays(day, 1);
  return day;
}

function previousTradingDay(value: string, symbol = "") {
  let day = value;
  while (!isTradingDay(day, symbol)) day = addDays(day, -1);
  return day;
}

function parsePublishedAt(value: unknown) {
  const raw = String(value || "");
  if (/^\d{8}T\d{6}$/.test(raw)) {
    return new Date(`${raw.slice(0, 4)}-${raw.slice(4, 6)}-${raw.slice(6, 8)}T${raw.slice(9, 11)}:${raw.slice(11, 13)}:${raw.slice(13, 15)}Z`);
  }
  return new Date(raw);
}

function marketParts(value: Date, symbol = "") {
  const korean = symbol.endsWith(".KS");
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: korean ? "Asia/Seoul" : "America/New_York",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    hourCycle: "h23",
  }).formatToParts(value);
  const part = (type: Intl.DateTimeFormatPartTypes) => parts.find((item) => item.type === type)?.value || "";
  return { date: `${part("year")}-${part("month")}-${part("day")}`, hour: Number(part("hour")) };
}

export function expectedReactionDate(publishedAt: unknown, symbol = "") {
  const parsed = parsePublishedAt(publishedAt);
  if (Number.isNaN(parsed.getTime())) return "";
  const local = marketParts(parsed, symbol);
  if (!isTradingDay(local.date, symbol)) return nextTradingDay(local.date, symbol);
  const closeHour = symbol.endsWith(".KS") ? 15 : 16;
  return local.hour >= closeHour ? nextTradingDay(addDays(local.date, 1), symbol) : local.date;
}

export function latestCompletedTradingDay(now = new Date(), symbol = "") {
  const local = marketParts(now, symbol);
  if (isTradingDay(local.date, symbol) && local.hour >= 17) return local.date;
  return previousTradingDay(addDays(local.date, -1), symbol);
}

export function marketImpactNeedsRefresh(row: Row, now = new Date()) {
  const impacts = asList(row.market_impact) as MarketImpact[];
  const candidates = impacts.length ? impacts : impactSymbols(row).map((symbol) => ({ symbol }));
  return candidates.some((impact) => {
    const expected = expectedReactionDate(row.published_at, impact.symbol);
    const due = expected && expected <= latestCompletedTradingDay(now, impact.symbol);
    return Boolean(due && (impact.pending || impact.error || impact.reaction_pct == null));
  });
}

export function calculateMarketImpact(symbol: string, publishedAt: unknown, bars: PriceBar[]): MarketImpact {
  const expected = expectedReactionDate(publishedAt, symbol);
  const reactionIndex = bars.findIndex((bar) => bar.date >= expected);
  const previous = reactionIndex > 0 ? bars[reactionIndex - 1] : undefined;
  const reaction = reactionIndex >= 0 ? bars[reactionIndex] : undefined;
  if (!previous) return { symbol, error: "missing price window" };
  if (!reaction) {
    return {
      symbol,
      pending: "waiting_for_reaction_close",
      expected_reaction_date: expected,
      prev_date: previous.date,
      prev_close: round(previous.close, 4),
    };
  }
  const next = bars[reactionIndex + 1];
  const third = bars[reactionIndex + 3];
  return {
    symbol,
    prev_date: previous.date,
    reaction_trade_date: reaction.date,
    prev_close: round(previous.close, 4),
    reaction_close: round(reaction.close, 4),
    next_close: next ? round(next.close, 4) : null,
    reaction_pct: pct(previous.close, reaction.close),
    next_day_pct: next ? pct(reaction.close, next.close) : null,
    three_session_pct: third ? pct(previous.close, third.close) : null,
  };
}

async function fetchBars(symbol: string, publishedAt: unknown): Promise<PriceBar[]> {
  const event = parsePublishedAt(publishedAt);
  const start = new Date(event);
  const end = new Date(event);
  start.setUTCDate(start.getUTCDate() - 10);
  end.setUTCDate(end.getUTCDate() + 11);
  const period1 = Math.floor(start.getTime() / 1000);
  const period2 = Math.floor(end.getTime() / 1000);
  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(symbol)}?period1=${period1}&period2=${period2}&interval=1d&events=history`;
  const response = await fetch(url, { headers: { "user-agent": "ghost-monitor/1.0" }, signal: AbortSignal.timeout(12_000) });
  if (!response.ok) throw new Error(`price HTTP ${response.status}`);
  const payload = await response.json() as {
    chart?: { result?: Array<{ timestamp?: number[]; indicators?: { quote?: Array<{ close?: Array<number | null> }>; adjclose?: Array<{ adjclose?: Array<number | null> }> } }> };
  };
  const result = payload.chart?.result?.[0];
  if (!result) return [];
  const closes = result.indicators?.adjclose?.[0]?.adjclose || result.indicators?.quote?.[0]?.close || [];
  return (result.timestamp || []).flatMap((timestamp, index) => {
    const close = closes[index];
    return typeof close === "number" && Number.isFinite(close)
      ? [{ date: new Date(timestamp * 1000).toISOString().slice(0, 10), close }]
      : [];
  });
}

function impactSymbols(row: Row) {
  const direct = asList(row.symbols).map(String);
  const directions = row.ticker_directions && typeof row.ticker_directions === "object"
    ? Object.keys(row.ticker_directions as Record<string, unknown>) : [];
  return [...new Set([...direct, ...directions])].filter(Boolean).slice(0, 8);
}

export async function enrichMarketImpact(row: Row, now = new Date()) {
  if (!marketImpactNeedsRefresh(row, now)) return false;
  const symbols = impactSymbols(row);
  if (!symbols.length) return false;
  row.market_impact = await Promise.all(symbols.map(async (symbol) => {
    try {
      return calculateMarketImpact(symbol, row.published_at, await fetchBars(symbol, row.published_at));
    } catch (error) {
      return { symbol, error: error instanceof Error ? error.message : "price fetch failed" };
    }
  }));
  return true;
}
